exporter/awskinesisexporter/encoding/encoder.go

```diff
@@ -0,0 +1,36 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encoding
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+var (
+	// ErrUnsupportedEncodedType is used when the encoder type does not support the type of encoding
+	ErrUnsupportedEncodedType = errors.New("unsupported type to encode")
+)
+
+// Encoder allows for the internal types to be converted to an consumable
+// exported type which is written to the kinesis stream
+type Encoder interface {
```

Comment:
Encoder sounds like something that takes some data in, translates it to another format and returns the translated data but the implementation below also seems to be responsible for actually exporting data to kinesis. Can we rename this to something like `KinesisForwarder` or `KinesisIngester`.

Reply:
I am actually borrowing this pattern from (encoding/json.Encoder)[https://golang.org/pkg/encoding/json/#Encoder] where you pass the writer to object and pass the types you want to encode to the writer.

Doing this meant that I wasn't breaking current behavour that was originally set for the exporter while allowing for future definitions to be added (coming once this has been added back into the contrib).

`KinesisForwarder` and `KinesisIngester` are not a good fit because that implies that the encoding and the actual kinesis producer implementation are tightly coupled and could lead to some avoidable code duplication. Furthermore, it is also breaks the stutter naming rule in Go.

Would it be fair to suggest `EncodeWriter` as the interface name?

Reply:
> Probably that's what is causing the conflict in my head :) When I saw encoder, I immediately associated it with ro expect an io.Writer somewhere. Seeing an awskinesis.Exporter was a bit surpising. Probably not a big deal.

Super sorry for the confusion, I just really like the method signature of: `encoding.Jaeger(exporter).Encode(data)`.
Mainly because it is what I use to when dealing with JSON.

> But we are already doing that by taking a dependency on awskinesis.Exporter. So naming it like that would only reflect the reality IMO.

Mmm, not wrong here. To help clarify my intent with this approach, I wanted to help keep backwards compatibility with jaeger while allowing new types of encoding / translations to be used. I am not too keen on using words such as Forwarder or Ingester because both don't infer that data is being mutated. I would prefer to keep it along the lines of Encoder or Translator. 

I believe I have seen translator packages used for other packages that allow for converted data types being exported, would it be wise to follow that pattern?

> BTW do you expect to consume this package as a library somewhere else? If not, why not make the interface private? That way we can change it's name pretty easily in future if we feel the need. It'd also make this whole discussion a lot less important :)

No :D My intent is to keep this internal to the kinesis exporter since it is rather tightly coupled on how the collector works.

Reply:
OK. Let's make it internal then and we should be good.

>  I am not too keen on using words such as Forwarder or Ingester because both don't infer that data is being mutated. I would prefer to keep it along the lines of Encoder or Translator.

Is the data being mutated though? It is translated to another form. The original values are discarded by the exporter. I think most other exporters do the same. They convert internal data into the format they want to export in. They all call their internal implementations exporter or forwarder.

>  I just really like the method signature of: encoding.Jaeger(exporter).Encode(data).
Mainly because it is what I use to when dealing with JSON.

This is exactly what feels a bit strange to me. Looking this function call doesn't tell me that this involves a network operation. It looks like this thing is only tranlating bytes into another form. People will need to look at the implementation or read the docs to realize this is also sending the encoded data to kinesis.

That said, if we make it private, I can live with as we can change it later if we want to.

Reply:
Alrighty, I have gone moved the package to be internal with a few other renames.

Reply:
Sorry, what I meant was make them private in the same package by making sure the names start with lower case 😬 Sorry about the miscommunication. 

I don't mind if you want to keep them in internal now that you've done the work though but not sure that the guidelines are about adding stuff to it. I think generally we use it to host code used by multiple packages in this project but something we don't want to support as part of the public API when using this repo as a library.

Reply:
Aaaa, right! completely flew by me.

I think internal works because there is nothing external to this that would benefit from using the translate.ExportWriters besides this exporter.